morl_baselines/multi_policy/multi_policy_moqlearning/mp_mo_q_learning.py

```diff
@@ -107,48 +106,144 @@ def get_config(self) -> dict:
+            The action to take.
         """
-        discounted_rewards = []
-        rewards = []
-        for a in self.agents:
-            _, _, vec, discounted_vec = a.policy_eval(eval_env=self.env, weights=a.weights, writer=self.writer)
-            discounted_rewards.append(discounted_vec)
-            rewards.append(vec)
-        if self.log:
-            print(f"Evaluation of all agents: {rewards}")
-            print(f"discounted: {discounted_rewards}")
-        return rewards, discounted_rewards
-
-    def train(self):
-        """Train the algorithm."""
-        start_time = time.time()
-        training_epoch = int(self.num_timesteps / self.eval_freq)
-        for e in range(training_epoch):
-            print(f"Training epoch #{e}")
-            for a in self.agents:
-                a.train(
-                    start_time,
-                    total_timesteps=self.eval_freq,
-                    reset_num_timesteps=False,
+        q_vals = np.stack([policy.scalarized_q_values(state, w) for policy in self.policies])
+        _, action = np.unravel_index(np.argmax(q_vals), q_vals.shape)
+        return int(action)
+
+    @override
+    def eval(self, obs: np.array, w: Optional[np.ndarray] = None, policy_ind: Optional[int] = None) -> int:
+        """If use_gpi is True, return the action given by the GPI policy. Otherwise, return the action given by the policy with the given index."""
+        if self.use_gpi:
+            return self._gpi_action(obs, w)
+        else:
+            return self.policies[policy_ind].eval(obs, w)
+
+    def delete_policies(self, delete_indx: List[int]):
+        """Delete the policies with the given indices."""
+        for i in sorted(delete_indx, reverse=True):
+            self.policies.pop(i)
+
+    def train(
+        self,
+        num_iterations: int,
+        timesteps_per_iteration: int,
+        eval_freq: int = 1000,
+        eval_env=None,
+        ref_point: Optional[np.ndarray] = None,
+        test_weights: Optional[np.ndarray] = None,
+        num_episodes_eval: int = 10,
+    ):
+        """Learn a set of policies.
+
+        Args:
+            num_iterations: The number of iterations/policies to train.
+            timesteps_per_iteration: The number of timesteps per iteration.
+            eval_freq: The frequency of evaluation.
+            eval_env: The environment to use for evaluation.
+            ref_point: The reference point for the hypervolume calculation.
+            test_weights: The weight vectors to use for evaluation (e.g. for expected utility).
+            epsilon_linear_support: The epsilon value for the linear support algorithm.
+            num_episodes_eval: The number of episodes used to evaluate the value of a policy.
+        """
+        if eval_env is None:
+            eval_env = deepcopy(self.env)
+
+        if test_weights is None:
+            test_weights = equally_spaced_weights(self.reward_dim, n=64)
```

Comment:
Why 64, not 100 for example? How do you set that?

Reply:
We should discuss this as well. Should we require the user always to pass the test_weights? We also have to be careful when evaluating the algos to use the same random_seed when generating the test weights

Reply:
Here is how I did in the metrics PR:
https://github.com/LucasAlegre/morl-baselines/pull/36/files#diff-10c2886bd6c1dbe1a867825f85d9bb017e3df5ebf3731205aaad10210c4f71dbR231

Reply:
I think it is important to differentiate weights used for training and weights for eval. Other than that, how many weights to use in evaluation is like a question mark for me... :/